timemachine/fe/rbfe.py

```diff
@@ -0,0 +1,378 @@
+    )
+
+    steps_per_frame = 750
+    # production
+    n_steps = n_frames * steps_per_frame
+    all_nrgs, all_coords, all_boxes = ctxt.multiple_steps_U(
+        lamb=lamb,
+        n_steps=n_steps,
+        lambda_windows=[lamb],
+        store_u_interval=steps_per_frame,
+        store_x_interval=steps_per_frame,
+    )
+
+    assert all_coords.shape[0] == n_frames
+    assert all_boxes.shape[0] == n_frames
+
+    return all_coords, all_boxes
+
+
+def get_batch_U_fns(bps, lamb):
+    # return a function that takes in coords, boxes, lambda
+    all_U_fns = []
+    for bp in bps:
+
+        def batch_U_fn(xs, boxes, bp_impl):
+            Us = []
+            for x, box in zip(xs, boxes):
+                # tbd optimize to "selective" later
+                _, _, U = bp_impl.execute(x, box, lamb)
+                Us.append(U)
+            return np.array(Us)
+
+        # extra functools.partial is needed to deal with closure jank
+        all_U_fns.append(functools.partial(batch_U_fn, bp_impl=bp))
+
+    return all_U_fns
+
+
+class HostConfig:
+    def __init__(self, omm_system, conf, box):
+        self.omm_system = omm_system
+        self.conf = conf
+        self.box = box
+
+
+def generate_samples(st, host_config, lambda_schedule, n_frames):
+    """
+    Generate samples given a lambda_schedule. Let L be the number
+    of lambda windows T be the number of frames.
+
+    Parameters
+    ----------
+    st: SingleTopology
+        Topology container for the alchemical ligand
+
+    host_config: HostConfig
+        Configuration of the host
+
+    lambda_schedule: np.array(L,)
+        Sequence of lambda windows
+
+    n_frames: int
+        Number of samples to generate per window
+
+    Returns
+    -------
+    3-tuple of all_frames, all_boxes, all_U_fns
+        all_frames has shape (L,T,N,3), all_boxes has shape (L,T,3,3), all_U_fns has shape (L,)
+
+    """
+    host_bps, host_masses = openmm_deserializer.deserialize_system(host_config.omm_system, cutoff=1.2)
+    host_conf = minimizer.minimize_host_4d(
+        [st.mol_a, st.mol_b],
+        host_config.omm_system,
+        host_config.conf,
+        st.ff,
+        host_config.box,
+    )
+
+    all_U_fns = []
+    all_frames = []
+    all_boxes = []
+
+    for lamb in lambda_schedule:
+        hgs = st.combine_with_host(convert_bps_into_system(host_bps), lamb=lamb)
+        all_U_fns.append(hgs.get_U_fns())
+        bound_impls = [p.bound_impl(np.float32) for p in hgs.get_U_fns()]
+        # minimize water box around the ligand by 4D-decoupling
+        mol_a_conf = get_romol_conf(st.mol_a)
+        mol_b_conf = get_romol_conf(st.mol_b)
+        ligand_conf = st.combine_confs(mol_a_conf, mol_b_conf)
+        combined_conf = np.concatenate([host_conf, ligand_conf])
+        x0 = combined_conf
+        v0 = np.zeros_like(x0)
+        box0 = host_config.box
+        group_idxs = get_group_indices(get_bond_list(hgs.bond))
+        seed = np.random.randint(0, 1000)
```

Comment:
is there a reason for this (vs just passing a fixed seed?)

Reply:
was a temporary solution - wasn't sure if i should have the user put in a single seed, (then increment it by # lambda windows along the way) or have them pass in an arrya of seeds.

Reply:
fixed by propagating seeds all the out